### PR TITLE
Filter all products block by attribute

### PR DESCRIPTION
**Pull request by mikejolley**
Introduces an attribute filter block for use alongside the all products block. 

Ref: #713 Note some acceptance criteria in that issue are not met in this MVP.

TODO:
- [x] Convert to store namespace
- [x] The list should be truncated if it contains more than 10 attributes
- [x] Add attribute selector to inspector sidebar as per the mocks
- [x] Discuss headings and 'toggle' in acceptance criteria

### Screenshots

Screenshot 2019-11-05 at 15 23 06

### How to test the changes in this Pull Request:

The best way to test these in conjunction with other filters is as follows:

1.  Create a column block with 2 columns
2. Add the all products block to one column
3. Add a price slider, and 2 attribute filter blocks to the other column. Setup one to use attribute 'size' and one to use 'color'. This is assuming you are using woo core sample data. You can use heading blocks to label them if you wish.

Then you can test how the filters work on the frontend.

### Changelog

> Introduce a filter products by attribute block which works alongside the new "All products" block.

### CODE REVIEW

File: assets/js/blocks/attribute-filter/edit.js [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+						{ __( 'Done', 'woo-gutenberg-products-block' ) }
+					</Button>
+				</div>
+			</Placeholder>
+		);
+	};
+
+	return (
+		<Fragment>
+			{ 0 === ATTRIBUTES.length ? (
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/edit.js]: There is something weird here, `ATTRIBUTES` is an object with weird keys, for me, so this check is always false:
image
Can you reproduce that too?

File: assets/js/blocks/attribute-filter/edit.js [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+					'woo-gutenberg-products-block'
+				) }
+			</p>
+			<Button
+				className="wc-block-attribute-filter__add_attribute_button"
+				isDefault
+				isLarge
+				href={
+					ADMIN_URL +
+					'edit.php?post_type=product&page=product_attributes'
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/edit.js]: We could use `getAdminLink` from `@woocommerce/navigation` here:
```JS
getAdminLink( 'edit.php?post_type=product&page=product_attributes' )`
```

File: assets/js/blocks/attribute-filter/frontend.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		};
+		el.classList.remove( 'is-loading' );
+
+		if ( ! attributes.attributeId ) {
+			return null;
+		}
+
+		render( <Block attributes={ attributes } />, el );
+	} );
+}
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/frontend.js]: Could we use the [`renderFrontend`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/utils/render-frontend.js) function here?

File: assets/js/blocks/attribute-filter/index.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		},
+	},
+	edit,
+	/**
+	 * Save the props to post content.
+	 */
+	save( { attributes } ) {
+		const { showCounts, displayStyle, queryType, attributeId } = attributes;
+		const data = {
+			'data-attributeid': attributeId,
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/index.js]: Is there any preference on using `data-attributeid` instead of `data-attribute-id`? Same for the attributes below.

File: assets/js/blocks/attribute-filter/block.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	useEffect( () => {
+		setCurrentAttribute(
+			find( ATTRIBUTES, [ 'attribute_id', attributeId + '' ] )
+		);
+	}, [ attributeId ] );
+
+	useEffect( () => {
+		const taxonomy = 'pa_' + currentAttribute.attribute_name;
+
+		if ( ! taxonomy ) {
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/block.js]: `taxonomy` will always resolve to true here, no?

File: assets/js/blocks/attribute-filter/block.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+			return;
+		}
+
+		const newProductAttributes = productAttributes.filter(
+			( item ) => item.attribute !== taxonomy
+		);
+		const slug = join( checkedOptions, ',' );
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/block.js]: Could we use the native method here?
```suggestion
		const slug = checkedOptions.join( ',' );
```

File: assets/js/blocks/attribute-filter/block.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+		filteredCounts,
+		allTerms,
+		showCounts,
+		filteredCountsIsLoading,
+		allTermsIsLoading,
+	] );
+
+	useEffect( () => {
+		setCurrentAttribute(
+			find( ATTRIBUTES, [ 'attribute_id', attributeId + '' ] )
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/block.js]: Why `+ ''`?

File: assets/js/blocks/attribute-filter/edit.js [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+	Placeholder,
+	Disabled,
+	PanelBody,
+	ToggleControl,
+	Button,
+	Toolbar,
+	withSpokenMessages,
+} from '@wordpress/components';
+import Gridicon from 'gridicons';
+import { ATTRIBUTES } from '@woocommerce/block-settings';
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/edit.js]: This should go to `Internal dependencies` too.

File: assets/js/blocks/attribute-filter/block.js (hunk 7 above)
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/block.js]: Casts the number to a string which the setting returns.

Comment (Aljullu) [reply assets/js/blocks/attribute-filter/block.js]: Makes sense, thank you. :+1: 

In these cases, I find `attributeId.toString()` is easier to understand and makes clear it's not a bug, but I'm fine leaving it as-is if I'm the only one.

File: assets/js/blocks/attribute-filter/index.js (hunk 4 above)
Comment (nerrad) [reply assets/js/blocks/attribute-filter/index.js]: I'd prefer hyphenating all words.  That will result in clearer camelcase properties too.

File: assets/js/blocks/attribute-filter/block.js [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+ * External dependencies
+ */
+import {
+	useCollection,
+	useQueryStateByKey,
+	useQueryStateContext,
+} from '@woocommerce/base-hooks';
+import { useCallback, Fragment, useEffect, useState } from '@wordpress/element';
+import { find, join, sortBy } from 'lodash';
+import { ATTRIBUTES } from '@woocommerce/block-settings';
```
[the remark this answers is not in the document]
Comment (Aljullu) [reply assets/js/blocks/attribute-filter/block.js]: Ok, sounds as a good compromise to me. Will mark this as solved, then.

File: assets/js/base/components/checkbox-list/index.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+			const checkedValue = event.target.value;
+			const newChecked = checked.filter(
+				( value ) => value !== checkedValue
+			);
+
+			if ( isChecked ) {
+				newChecked.push( checkedValue );
+			}
+
+			newChecked.sort();
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: @nerrad This is a checkbox change event so presumably this callback only fires when values change, meaning the state must update. The only edge case would be someone manually triggering change events but that seems unlikely.

Good call on the sort.

File: assets/js/base/components/checkbox-list/style.scss [hunk 11]
```diff
@@ -0,0 +1,17 @@
+.editor-styles-wrapper .wc-block-checkbox-list,
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/style.scss]: Greater specificity is required in the editor to avoid editor styles on list elements (margins, padding, list style).

File: assets/js/base/components/checkbox-list/index.js [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+		[ options, checked ]
+	);
+
+	const renderOptions = useCallback( () => {
+		// Truncate options if > the limit + 5.
+		const shouldTruncateOptions = options.length > limit + 5;
+		const showOptions =
+			shouldTruncateOptions && ! showExpanded
+				? [ ...options.slice( 0, limit ) ]
+				: options;
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: If you can tab to the button and it's labelled correctly, probably. Hidden elements (with CSS) is no more accessible - screen readers still do not see them.

File: assets/js/base/components/checkbox-list/index.js [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+						>
+							{ showExpanded
+								? __(
+										'Show less',
+										'woo-gutenberg-products-block'
+								  )
+								: __(
+										'Show more',
+										'woo-gutenberg-products-block'
+								  ) }
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: Needs aria text

File: assets/js/blocks/attribute-filter/block.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/block.js]: Attributes is something I consider a setting that doesn't change often, and I think getting this from the server will be more performant that more API requests. It's related to the users woo environment... like post counts or other settings we pull here.

[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/blocks/attribute-filter/block.js]: What potential is there for the size of the attributes data package?  Is it typical for it to be a smaller number of bytes or larger?  Reason I ask is because I wonder if it could impact time to first paint of a site loading by virtue of increasing the html package for large data sets.  That's where lazy loading of attributes might have more of an impact.  Granted, if the typical number of attributes is fairly small, this would definitely be an over-optimization.

I'm fine with not addressing this right now.

[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/block.js]: @nerrad Attributes (these are taxonomies) can vary in size and is unknown, however, woo core already caches these in a transient and they are loaded on each page anyway (to register them as taxonomies), so there is no negative to using these as block settings as it's essentially 'free' vs an API request.

File: assets/js/blocks/attribute-filter/block.js [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+	useEffect( () => {
+		const newProductAttributes = productAttributes.filter(
+			( item ) => item.attribute !== taxonomy
+		);
+		const slug = checkedOptions.join( ',' );
+
+		if ( slug ) {
+			const updatedQuery = {
+				attribute: taxonomy,
+				operator: 'or' === queryType ? 'in' : 'and',
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/block.js]: This is a yoda condition, right? :slightly_smiling_face: 
```suggestion
				operator: queryType === 'or' ? 'in' : 'and',
```

File: assets/js/blocks/attribute-filter/block.js [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+		setProductAttributes( sortBy( newProductAttributes, 'attribute' ) );
+	}, [ checkedOptions, taxonomy, productAttributes ] );
+
+	const onChange = useCallback( ( checked ) => {
+		setCheckedOptions( checked );
+	}, [] );
+
+	if ( ! taxonomy ) {
+		return null;
+	}
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/block.js]: Could we move this check closer to the top of the functional component? If we are returning `null`, I guess it's better to return as early as possible to avoid useless computation.

File: assets/js/blocks/attribute-filter/edit.js [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+					<Button isDefault onClick={ onDone }>
+						{ __( 'Done', 'woo-gutenberg-products-block' ) }
+					</Button>
+				</div>
+			</Placeholder>
+		);
+	};
+
+	return (
+		<Fragment>
```
Comment (Aljullu) [inline assets/js/blocks/attribute-filter/edit.js]: I think this `<Fragment>` can be removed from here too:
```JSX

	return 0 === Object.keys( ATTRIBUTES ).length ? (
		noAttributesPlaceholder()
	) : (
		...
	);
```

File: assets/js/base/components/checkbox-list/index.js [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+} ) => {
+	// Holds all checked options.
+	const [ checked, setChecked ] = useState( [] );
+	const [ showExpanded, setShowExpanded ] = useState( false );
+
+	useEffect( () => {
+		onChange( checked );
+	}, [ checked ] );
+
+	const placeholder = useMemo( () => {
```
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: As I mentioned in my previous comment:

> Advantage of using useMemo here is that it only calculates the value once when this component is mounted. Thus any additional calls to the function on lifecycle changes while the component is mounted will return the memoized value instead of regenerating the list items.

Which would you prefer?

No `useMemo`:

Image 2019-11-07 at 11 43 03 AM

With `useMemo`:

Image 2019-11-07 at 11 43 51 AM

I added `console.log( 'calculating...' )` whenever placeholder is invoked for the above.

File: assets/js/blocks/attribute-filter/block.js (hunk 15 above)
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/block.js]: Conditionals must be after hooks.

File: assets/js/base/components/checkbox-list/index.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: Yeah I was coming back to say this; it re-renders the placeholder many times. `useMemo` does not so stick with that.

[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: Incidentally, understanding what is happening in react with functional components (and now with Hooks) was one of the hardest things for me to wrap my head around (and even now I still sometimes forget and get it wrong). When in doubt, I'll drop in some console.logs and watch what happens :)

[the remark this answers is not in the document]
Comment (Aljullu) [reply assets/js/base/components/checkbox-list/index.js]: You're right, I'm new to hooks so it's still not clear to me when it's worth using `useMemo()` and when it's not. Technically, we could also wrap the [classes assignment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1127/files#diff-53d339601cf4e1728e109362c7ca6207R159) into a `useMemo()` function, right? But that would probably be over-kill since it's not an expensive function at all. Creating the `placeholder` seems a bit more expensive, but when is a function "expensive enough" to require `useMemo()`? I wonder if you have a rule of thumb to know when it's worth using `useMemo()` and when it's not.

Regarding this specific case, why not moving `placeholder`'s definition outside of the `CheckboxList` component so it's only computed once without the need of `useMemo()`? There might still be the edge case in which we import `chebox-list/index.js` but don't create `CheckboxList`, in that case we would be creating the `placeholder` but not using its value.  But I think we can neglect this case.

[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: > Technically, we could also wrap the classes assignment into a useMemo() function, right?

Right.  But as you pointed out, it's overkill because it's not expensive.

> but when is a function "expensive enough" to require useMemo()? I wonder if you have a rule of thumb to know when it's worth using useMemo() and when it's not.

No rule of thumb, and likely the performance gain with adding `useMemo` here is not huge.  But for things like a placeholder list I think it doesn't hurt.

> Regarding this specific case, why not moving placeholder's definition outside of the CheckboxList component so it's only computed once without the need of useMemo()?

Yup, anytime there are no dependencies, it's a good candidate to simply define it as a variable outside of the function. But there's also advantages to keeping things defined in the component using them when it comes to code review, and context.

File: assets/js/base/components/checkbox-list/index.js [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+						</li>
+						{ shouldTruncateOptions &&
+							index === limit - 1 &&
+							renderedShowMore }
+					</Fragment>
+				) ) }
+				{ shouldTruncateOptions && renderedShowLess }
+			</Fragment>
+		);
+	}, [ options, checked, showExpanded, limit, onCheckboxChange ] );
```
Comment (nerrad) [inline assets/js/base/components/checkbox-list/index.js]: `renderedShowLess` and `renderedShowMore` should be dependencies

File: assets/js/blocks/attribute-filter/block.js [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+		( name, count ) => {
+			return (
+				<Fragment>
+					{ name }
+					{ showCounts && count && (
+						<span className="wc-block-attribute-filter-list-count">
+							{ count }
+						</span>
+					) }
+				</Fragment>
```
Comment (nerrad) [inline assets/js/blocks/attribute-filter/block.js]: Should probably have a key on the fragment here since this is part of a list.

File: assets/js/blocks/attribute-filter/block.js [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+			const updatedQuery = {
+				attribute: taxonomy,
+				operator: queryType === 'or' ? 'in' : 'and',
+				slug,
+			};
+			newProductAttributes.push( updatedQuery );
+		}
+
+		setProductAttributes( sortBy( newProductAttributes, 'attribute' ) );
+	}, [ checkedOptions, taxonomy, productAttributes ] );
```
Comment (nerrad) [inline assets/js/blocks/attribute-filter/block.js]: Need `queryType` added to the dependencies.

File: assets/js/base/components/checkbox-list/index.js [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+	}, [ showExpanded ] );
+
+	const renderedOptions = useMemo( () => {
+		// Truncate options if > the limit + 5.
+		const optionCount = options.length;
+		const shouldTruncateOptions = optionCount > limit + 5;
+		return (
+			<Fragment>
+				{ options.map( ( option, index ) => (
+					<Fragment key={ option.key }>
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: This fragment wraps 2 sibling `<li>` elements so is needed here?

File: assets/js/blocks/attribute-filter/frontend.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+ * Internal dependencies
+ */
+import Block from './block.js';
+
+const getProps = ( el ) => {
+	return {
+		attributes: {
+			attributeId: parseInt( el.dataset.attributeId || 0, 10 ),
+			showCounts: el.dataset.showCounts === 'true',
+			queryType: el.dataset.queryType,
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/frontend.js]: Defaults are set on block registration for these attributes. See the index file:

```
attributes: {
		attributeId: {
			type: 'number',
			default: 0,
		},
		showCounts: {
			type: 'boolean',
			default: true,
		},
		queryType: {
			type: 'string',
			default: 'or',
		},
	},
```

Also a pattern across all existing blocks?

File: assets/js/blocks/attribute-filter/block.js [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+
+	const [ queryState ] = useQueryStateByContext( 'product-grid' );
+	const [ productAttributes, setProductAttributes ] = useQueryStateByKey(
+		'product-grid',
+		'attributes',
+		[]
+	);
+
+	const filteredCountsQueryState = useMemo( () => {
+		// If doing an "AND" query, we need to remove current taxonomy query so counts are not affected.
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/block.js]: Without a taxonomy this query will be broken... I think the change to the useCollection hook would be good so we can avoid the query altogether if missing (since we cannot add a conditional before the hooks). Follow up?

File: assets/js/base/components/checkbox-list/index.js [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+			)
+		);
+	}, [ showExpanded ] );
+
+	const renderedOptions = useMemo( () => {
+		// Truncate options if > the limit + 5.
+		const optionCount = options.length;
+		const shouldTruncateOptions = optionCount > limit + 5;
+		return (
+			<Fragment>
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: Like the other mentioned above, this wraps sibling elements.

File: assets/js/base/components/checkbox-list/index.js [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+import './style.scss';
+
+/**
+ * Component used to show a list of checkboxes in a group.
+ */
+const CheckboxList = ( {
+	className,
+	onChange = () => {},
+	options = [],
+	isLoading = false,
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: That would also force people to pass `isLoading` in which may not be appropriate for all use case. Looking at the block, `isLoading` would be true when calling this component for the first time so there won't be flicker.

File: assets/js/base/components/checkbox-list/index.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: You're right there is _one_ Fragment that could remain.  However there's a couple things:

1. The inner level fragment should still not be needed.
2. The key should be on the li items, not the Fragment.  The problem with the Fragment having the key is that the "show more options" toggle will get re-rendered everytime the inner list re-renders even if it's not necessary.

The above points might not see important in the context of this particular usage, but it's pattern that could be more problematic in other contexts if it becomes a habit :). So I'm just promoting best practices here.

This code could get changed to:

```jsx
useMemo( () => {
	// Truncate options if > the limit + 5.
	const optionCount = options.length;
	const shouldTruncateOptions = optionCount > limit + 5;
	return (
		<Fragment>
			{ options.map( ( option, index ) => (
				<li
					key={ option.key }
					{ ...shouldTruncateOptions &&
						! showExpanded &&
						index >= limit && { hidden: true } }
				>
					<input
						type="checkbox"
						id={ option.key }
						value={ option.key }
						onChange={ onCheckboxChange }
						checked={ checked.includes( option.key ) }
					/>
					<label htmlFor={ option.key }>{ option.label }</label>
				</li>
			) ) }
			{ shouldTruncateOptions &&
				options.length > limit &&
				renderedShowMore }
			{ shouldTruncateOptions && renderedShowLess }
		</Fragment>
	);
}, [
	options,
	checked,
	showExpanded,
	limit,
	onCheckboxChange,
	renderedShowLess,
	renderedShowMore,
] );
```

Even better though, would be to handle the `renderedShowMore` and `renderedShowLess` outside of this memoized function.  I'd combine the `renderedShowMore` and `renderedShowLess` into one memoized function that spits out either of those or null based on the conditions. Then you could get rid of the last unnecessary fragment and end up with something like this:

```jsx
<ul className={ classes }>
    { isLoading ? placeholder : renderedOptions  }
    { renderedToggleOptions }
</ul>
```

File: assets/js/blocks/attribute-filter/block.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/blocks/attribute-filter/block.js]: + to follow up.

File: assets/js/base/components/checkbox-list/index.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: My reply above relates to this as well.

File: assets/js/blocks/attribute-filter/block.js [hunk 26]
```diff
@@ -26,7 +26,7 @@ import { getTaxonomyFromAttributeId } from '../../utils/attributes';
  * Component displaying an attribute filter.
  */
 const AttributeFilterBlock = ( { attributes } ) => {
-	const [ options, setOptions ] = useState( [] );
+	const [ options, setOptions ] = useState( null );
```
Comment (nerrad) [inline assets/js/blocks/attribute-filter/block.js]: hmm, why did you change this to null?  Have you verified any place assuming options is an array is checking it is an array first?

File: assets/js/blocks/attribute-filter/block.js [hunk 27]
```diff
@@ -170,7 +176,7 @@ const AttributeFilterBlock = ( { attributes } ) => {
 		<div className="wc-block-attribute-filter">
 			<CheckboxList
 				className={ 'wc-block-attribute-filter-list' }
-				options={ options }
+				options={ options ? options : [] }
```
Comment (nerrad) [inline assets/js/blocks/attribute-filter/block.js]: On problem with this construct is that you're always passing a NEW array along if `options` is a non truthy value.  So every call to this component by react to determine whether to re-render or not is going to trigger a re-render if nothing else changed (because we have a new array instance):

```js
render1Options = [];
render2Options = [];
render1Options !== render2Options
```

I'm not sure why you switched that anyways, as far as I can tell, options will always be an array?

File: assets/js/base/components/checkbox-list/index.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: Hmm but by having the show more after the list items, if you tab through (with keyboard) when it expands you're at the bottom of the list  rather than from the point of expansion.. that was my justification for having it appear inline. Originally I did it similar to you but only rendering visible list items. You were against that particular solution also.

[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: 🤔 hmm... I see your point.  I tried both behaviours and what you have definitely works better from an a11y pov. Let's roll with it.

[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/checkbox-list/index.js]: >  The problem with the Fragment having the key is that the "show more options" toggle will get re-rendered everytime the inner list re-renders even if it's not necessary.

Did we work around this by adding a key to the show more/show less li?

File: assets/js/blocks/attribute-filter/block.js (hunk 26 above)
Comment (mikejolley) [reply assets/js/blocks/attribute-filter/block.js]: Actually I abandoned a check I added this for so I will revert.

**Comment by mikejolley** [comment]
@nerrad I'll add tests separately if we're happy with the API changes. R/e the useCollection follow-up, are you handling it?

File: assets/js/base/components/checkbox-list/index.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (nerrad) [reply assets/js/base/components/checkbox-list/index.js]: > Did we work around this by adding a key to the show more/show less li?

Hmm... I don't think so (not sure honestly), but regardless I think in this case any accessibility improvements make it worth potential extra rendering for the time being (until we can think of a better pattern).
